include/xtensor/xbuilder.hpp

```diff
@@ -556,5 +556,259 @@ namespace xt
+            {
+            }
+
+            inline value_type operator()() const
+            {
+                // special case when called without args (happens when printing)
+                return value_type();
+            }
+
+            template <class... Args>
+            inline value_type operator()(Args... args, std::size_t last_idx) const
+            {
+                return m_source(args..., m_shape_last - last_idx);
+            }
+
+            inline value_type operator[](xindex idx) const
+            {
+                idx.back() = m_shape_last - idx.back();
+                return m_source[idx];
+            }
+
+            template <class It>
+            inline value_type element(It first, It last) const
+            {
+                return operator[](xindex(first, last));
+            }
+
+        private:
+            CT m_source;
+            const size_type m_shape_last;
+        };
+
+        template <class CT, class Comp>
+        struct trilu_fn
+        {
+            using xexpression_type = std::decay_t<CT>;
+            using value_type = typename xexpression_type::value_type;
+            using signed_idx_type = long int;
+
+            trilu_fn(CT source, int k, Comp comp)
+                : m_source(source), m_k(k), m_comp(comp)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(It begin, It end) const
+            {
+                // have to cast to signed int otherwise -1 can lead to overflow
+                return m_comp(signed_idx_type(*begin) + m_k, signed_idx_type(*(begin + 1))) ? m_source.element(begin, end) : value_type(0);
+            }
+
+        private:
+            CT m_source;
+            const signed_idx_type m_k;
+            const Comp m_comp;
+        };
+    }
+
+    /**
+     * @function diagonal(const xexpression<T>& arr)
+     * @brief Returns the elements on the diagonal of arr
+     *
+     * @param arr the input array
+     *
+     * @return xexpression with values of the diagonal
+     */
+    template <class E>
+    inline auto diagonal(E&& arr)
+    {
+        using CT = xclosure_t<E>;
+        return detail::make_xgenerator(detail::fn_impl<detail::diagonal_fn<CT>>(detail::diagonal_fn<CT>(std::forward<E>(arr))),
+                                       {arr.shape()[0]});
+    }
+
+    /**
+     * @function diag(const xexpression<T>& arr)
+     * @brief xexpression with values of arr on the diagonal, zeroes otherwise
+     *
+     * @param arr the 1D input array of length n
+     *
+     * @return xexpression function with shape n x n and arr on the diagonal
+     */
+    template <class E>
+    inline auto diag(E&& arr)
+    {
+        using CT = xclosure_t<E>;
+        return detail::make_xgenerator(detail::fn_impl<detail::diag_fn<CT>>(detail::diag_fn<CT>(std::forward<E>(arr))),
+                                       {arr.shape()[0], arr.shape()[0]});
```

Comment:
You forward `arr` and then attempt to access `array.shape`. This could be the cause of the error you were seeing. It could be interested to try storing `shape()[0]` in a temporary variable first.